Rework `spack.database.InstallStatuses` into a flag

alalazo opened the pull request:
In `spack.database` we use a custom enum-like class `InstallStatuses` for legacy reasons[^1]. 

In this PR we turn that class into an `enum.Flag`, which has some small benefits:
- Flags are composable with `|`. That allows us to simplify the signature of a couple of methods
- We can easily define a constant in the enum representing "any" status, and stop using the builtin `any` for that purpose

On top of that:
- More type-hinting is added to query functions.
- In `spack find` the  `--only-missing` option is mutually exclusive with `--only-deprecated`. Both can't be mixed with `--missing` and `--deprecated`.

[^1]: This class was added when Spack still had to support Python 2.6
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

alalazo committed:
Type-hinted Database.query return value

alalazo committed:
Use an enum.Flag for the status query, add type-hints

Also, don't use the built-in any as a sentinel, but create an
appropriate value with flags.

alalazo committed:
Add more type-hinting, remove mentions of spack.database.InstallStatus

haampie commented on lib/spack/spack/cmd/find.py:
```diff
@@ -165,14 +166,23 @@ def setup_parser(subparser):
+        installed = StatusQuery.DEPRECATED
+
+    if args.missing:
+        installed |= StatusQuery.MISSING
```
So, the only combination that used to work is `--deprecated --missing`, any other combination of `--missing`, `--deprecated`, `--only-missing`, `--only-deprecated` fails?

This code allows `--missing --only-missing` iiuc. It's not identical. I guess nobody cares.

alalazo replied:
> Maybe make the whole set --{only-}{missing,deprecated} mutually exclusive then?

If we want `--only-missing --missing` to error (which is fine according to the help, I guess), I can just tweak the initial conditionals:
```python
if args.only_missing and (args.deprecated or args.missing):
    raise RuntimeError("cannot use --only-missing with --deprecated, or --missing")
```
Making them all mutually exclusive prevents `--deprecated --missing` which I think should be valid?

haampie replied:
yeah, would just leave the original branches

alalazo replied:
The original branches made little sense, as they allowed options together that should have been incompatible:
![Screenshot from 2024-10-30 14-47-05](https://github.com/user-attachments/assets/00b938ef-0d34-4f2a-b0c9-31bd94521d07)

Let me know if https://github.com/spack/spack/pull/47321/commits/10e54fd4cf637b9a5dad8313f358d3d2209f6e9f seems ok

alalazo committed:
Fix --only-missing --missing

alalazo committed:
Move the enum to a separate module

alalazo committed:
Rename StatusQuery -> DBStatusQuery

Since the enum flag has been moved into a separate module,
it makes sense to prefix it with "DB".

alalazo committed:
Further cleanup of imports

alalazo committed:
Shorten DBStatusQuery -> DBStatus

alalazo commented on lib/spack/spack/enums.py:
```diff
@@ -0,0 +1,17 @@
+import enum
+
+
+class DBStatus(enum.Flag):
```
This one was renamed, cause we have 3 different `class InstallStatus` in `develop` right now.

haampie replied:
It's not a status of the DB but of the install record / spec, and we never use an abbreviation like "DB" elsewhere. I'm not a fan of this name.

alalazo replied:
I can use `InstallRecordStatus`. Point is, I would like to avoid duplicate class names.

alalazo replied:
Done in https://github.com/spack/spack/pull/47321/commits/3008094c674f53a933b7d02f4b9ed0280ea0cbc1

haampie commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -252,7 +252,7 @@ def _associate_built_specs_with_mirror(self, cache_key, mirror_url):
             spec_list = [
                 s
-                for s in db.query_local(installed=any)
+                for s in db.query_local(installed=spack_db.ANY_STATUS)
```
wrong import

alalazo replied:
https://github.com/spack/spack/pull/47321/commits/8d82ac5b01f7f126ce620d3a3f9a9eee2b90fac6

haampie commented on lib/spack/spack/test/cmd/reindex.py:
```diff
@@ -5,7 +5,7 @@
 
 import spack.store
-from spack.database import Database
+from spack.database import ANY_STATUS, Database
```
wrong import

alalazo replied:
https://github.com/spack/spack/pull/47321/commits/8d82ac5b01f7f126ce620d3a3f9a9eee2b90fac6

alalazo committed:
Import from spack.enums

alalazo committed:
DBStatus -> InstallRecordStatus

alalazo committed:
ANY_STATUS -> InstallRecordStatus.ANY

haampie commented on lib/spack/spack/cmd/__init__.py:
```diff
@@ -267,39 +269,48 @@ def matching_specs_from_env(specs):
 
-def disambiguate_spec(spec, env, local=False, installed=True, first=False):
+def disambiguate_spec(
+    spec: "spack.spec.Spec",
```
can you drop all the `"`s? That shouldn't be necessary in cmd/*

alalazo replied:
Should be done

alalazo committed:
Remove quotes from "spack.spec.Spec" in `spack.cmd`

alalazo committed:
Remove quotes from "ev.Environment" in `spack.cmd`
